fivox/cudaImageSource.hxx

```diff
@@ -0,0 +1,131 @@
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef FIVOX_CUDAIMAGESOURCE_HXX
+#define FIVOX_CUDAIMAGESOURCE_HXX
+
+#include "cuda/simpleLFP.h"
+#include "cudaImageSource.h"
+#include "volumeHandler.h"
+
+namespace fivox
+{
+template< typename TImage >
+CudaImageSource< TImage >::CudaImageSource()
+    : ImageSource< TImage >()
+{
+}
+
+template< typename TImage >
+void CudaImageSource< TImage >::GenerateData()
+{
+    auto image = Superclass::GetOutput();
+    image->Allocate();
+    image->FillBuffer( 0 );
+
+    auto source = Superclass::_eventSource;
+    float* posX;
+    float* posY;
+    float* posZ;
+    float* radii;
+    float* values;
+    const int fsize = source->getNumEvents() * sizeof(float);
+    source->load();
+
+    const auto& sizeVoxels = Superclass::getSizeInVoxel();
+    const size_t width = sizeVoxels.x();
+    const size_t height = sizeVoxels.y();
+    const size_t depth = sizeVoxels.z();
+
+    cuda::VolumeInfo volInfo;
+    volInfo.dimensions.x = width;
+    volInfo.dimensions.y = height;
+    volInfo.dimensions.z = depth;
+
+    const auto& volExtent( Superclass::getSizeInMicrometer( ));
+    const size_t volSize( std::ceil( Superclass::getSizeInVoxel().find_max( )));
+    const VolumeHandler volumeHandler( volSize, volExtent );
+    volInfo.voxelSize = volumeHandler.computeSpacing()[0];
+
+    const auto& origin = Superclass::getBoundingBox().getMin();
+    volInfo.origin.x = origin[0];
+    volInfo.origin.y = origin[1];
+    volInfo.origin.z = origin[2];
+
+    cuda::Parameters parameters;
+    parameters.numEvents = source->getNumEvents();
+    parameters.cutoff = source->getCutOffDistance();
+
+    // copy input from host to device
+    gpuErrchk( cudaMalloc( (void**)&posX, fsize ));
+    gpuErrchk( cudaMalloc( (void**)&posY, fsize ));
+    gpuErrchk( cudaMalloc( (void**)&posZ, fsize ));
+    gpuErrchk( cudaMalloc( (void**)&radii, fsize ));
+    gpuErrchk( cudaMalloc( (void**)&values, fsize ));
+    gpuErrchk( cudaMemcpy( posX, source->getPositionsX(), fsize,
+                           cudaMemcpyHostToDevice ));
+    gpuErrchk( cudaMemcpy( posY, source->getPositionsY(), fsize,
+                           cudaMemcpyHostToDevice ));
+    gpuErrchk( cudaMemcpy( posZ, source->getPositionsZ(), fsize,
+                           cudaMemcpyHostToDevice ));
+    gpuErrchk( cudaMemcpy( radii, source->getRadii(), fsize,
+                           cudaMemcpyHostToDevice ));
+    gpuErrchk( cudaMemcpy( values, source->getValues(), fsize,
+                           cudaMemcpyHostToDevice ));
+
+    const int numVoxels = width * height * depth;
+
+    float* cudaOutput;
+    gpuErrchk( cudaMalloc( (void**)&cudaOutput, numVoxels * sizeof(float) ));
+    const float elapsed = cuda::simpleLFP( posX, posY, posZ, radii, values,
+                                           parameters, volInfo, cudaOutput );
+    LBINFO << "CUDA elapsed time: " << elapsed << " ms" << std::endl;
```

Comment:
Not trivial, it depends on the strategy we want to use (e.g. do it by chunks; start copying the output as it gets computed; etc). To be defined.